### ISSUE TITLE
Beta-testing 'mini.pick'

### ISSUE THREAD

**Issue by echasnovski** [echasnovski/mini.nvim#513]
Please leave your feedback about new [mini.pick](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-pick.md) module here. Feel free to either add new comment or positively upvote existing one.

Some things I am interested to find out (obviously, besides bugs):

- Are configuration and function/picker names intuitive enough?
- Are default values of settings convenient for you?
- Is documentation and examples clear enough?
- Does default matching algorithm give relevant enough results?

Thanks!

**Comment by Wansmer** [comment]
Hi, thanks for the cool plugin!

I'm trying to make a picker similar to `telescope.pickers.current_buffer_fuzzy_find({ default_text = vim.fn.expand('<cword>') })`, but failed.

How can I pass the word under the cursor as a query when creating the picker?

I tried using `MiniPick.set_picker_query()`, but it has no effect. I called both before `MiniPick.start()` and after.

Here is an example of what I tried to do:

```lua
    local pick = require('mini.pick')

    pick.registry.grep_live_current_buf = function()
      local lines = vim.api.nvim_buf_get_lines(0, 1, -1, false)
      local items = {}

      local path = vim.fn.expand('%p')
      for lnum, line in ipairs(lines) do
        items[lnum] = { text = line, lnum = lnum + 1, path = path }
      end

      pick.start({
        source = {
          items = items,
          name = 'Grep current buffer',
        },
      })
      pick.set_picker_query(vim.split(vim.fn.expand('<cword>'), ''))
    end
```

As I understood from the plugin code, the initial query is always set empty in the private function `H.picker_new`. Is it possible to add an option in `opts.source` to specify the initial query?

**Comment by echasnovski** [comment]
> I'm trying to make a picker similar to `telescope.pickers.current_buffer_fuzzy_find({ default_text = vim.fn.expand('<cword>') })`, but failed.

If it wasn't for only current buffer restriction, I'd suggest using `grep` directly. There is even [example of that in the help](https://github.com/echasnovski/mini.nvim/blob/32ccb0c17cfe7e4defb0e8d5328fd0e777dae93d/doc/mini-pick.txt#L653) (although, not in a very visible place). 

> How can I pass the word under the cursor as a query when creating the picker?

Your code is good, it only misses one important design: although `start()` is a non-blocking (allows other things to be executed), it is not fully asynchronous (it waits for user to choose and then returns certain value). It is documented both in ["Implementation details" of overview](https://github.com/echasnovski/mini.nvim/blob/32ccb0c17cfe7e4defb0e8d5328fd0e777dae93d/doc/mini-pick.txt#L196-L198) and (not entirely directly) in [`start()` itself](https://github.com/echasnovski/mini.nvim/blob/32ccb0c17cfe7e4defb0e8d5328fd0e777dae93d/doc/mini-pick.txt#L812).

So in your code by the time `set_picker_query()` is executed, picker is already stopped and thus it fails silently (as it is allowed to be called without active picker). If you call it directly before `start()`, it would behave the same (as there is no active picker *yet*). The solution here is to use `vim.schedule()` prior to `start()`.

Here is my attempt:
```lua
MiniPick.registry.buffer_lines = function(local_opts)
  -- Parse options
  local_opts = vim.tbl_deep_extend('force', { buf_id = nil, prompt = '' }, local_opts or {})
  local buf_id, prompt = local_opts.buf_id, local_opts.prompt
  local_opts.buf_id, local_opts.prompt = nil, nil

  -- Construct items
  if buf_id == nil or buf_id == 0 then buf_id = vim.api.nvim_get_current_buf() end
  local lines = vim.api.nvim_buf_get_lines(buf_id, 0, -1, false)
  local items = {}
  for i, l in ipairs(lines) do
    items[i] = { text = string.format('%d:%s', i, l), bufnr = buf_id, lnum = i }
  end

  -- Start picker while scheduling setting the query
  vim.schedule(function() MiniPick.set_picker_query(vim.split(prompt, '')) end)
  MiniPick.start({ source = { items = items, name = 'Buffer lines' } })
end
```

**Comment by Wansmer** [comment]
> So in your code by the time `set_picker_query()` is executed, picker is already stopped and thus it fails silently (as it is allowed to be called without active picker). If you call it directly before `start()`, it would behave the same (as there is no active picker _yet_). The solution here is to use `vim.schedule()` prior to `start()`.
> 
> Here is my attempt:

Thanks for the quick reply!

Yes, I tried using `vim.schedule` before, but that didn't help either. 
I copied and ran your code, but it doesn't seem to give the expected result either.

In the video, promt is expected to have a word under the cursor, but the prompt is empty. I also tried the example from the documentation, but in that case '<cword>' is also not passed and the result is empty:


https://github.com/echasnovski/mini.nvim/assets/46977173/49833e2a-3428-4bc3-8750-fdc9bc4f4136

**Comment by BeauSLM** [comment]
I saw this issue after I filed mine, huge apologies. TLDR is Pick grep searches .git folder, which isn't really useful. Excellent plugin!

#516

**Comment by echasnovski** [comment]
> I saw this issue after I filed mine, huge apologies. TLDR is Pick grep searches .git folder, which isn't really useful. Excellent plugin!
> 
> #516

No worries. I'll think about it.

**Comment by assistcontrol** [comment]
Gosh this one is amazing! The speed increase compared to Telescope is staggering. You should be really proud of what you've created here.

I'm struggling to figure out how to browse files in another directory. With Telescope, I had to `lcd` first, otherwise each entry was `../foo/bar/...` etc. Between the readme and the doc, I can't figure out the incantation here.

Is there a mechanism to tell `builtin.files()` to use a specific dir? If not, I'd definitely urge one as it's something that Telescope is sorely missing. `MiniPick.builtin.files({dir = '../foo/bar'})` would be huge for me.

**Comment by echasnovski** [comment]
> Gosh this one is amazing! The speed increase compared to Telescope is staggering. You should be really proud of what you've created here.

:heart: 

> I'm struggling to figure out how to browse files in another directory. With Telescope, I had to `lcd` first, otherwise each entry was `../foo/bar/...` etc. Between the readme and the doc, I can't figure out the incantation here.
> 
> Is there a mechanism to tell `builtin.files()` to use a specific dir? If not, I'd definitely urge one as it's something that Telescope is sorely missing. `MiniPick.builtin.files({dir = '../foo/bar'})` would be huge for me.

It is possible, as every source can have "current working directory". So with `builtin.files()` you can do `MiniPick.builtin.files(nil, { source = { cwd = '../foo/bar' } })`. If you plan to do that often interactively, you can [modify `files` in registry](https://github.com/echasnovski/mini.nvim/blob/0f93724569ebf27b6b3bc7130d8044a9020a8cc5/doc/mini-pick.txt#L1135-L1140) and use `:Pick files cwd='../foo/bar'`.

It is made a part of the source for it to be more persistent and not depend on current working directory. Allowing it also as the local option for built-in picker is a code duplication I would like to avoid. Thus the modification of `registry` approach.

------

To be honest, I did not realize that picking something from different directory is a common thing. I'd make it more prominent in the documentation.

Maybe this is a workflow thing. I use [`setup_auto_root()`](https://github.com/echasnovski/mini.nvim/blob/0f93724569ebf27b6b3bc7130d8044a9020a8cc5/doc/mini-misc.txt#L119) and can't really remember when I needed to pick something from outside of the root. If I need something, then I  usually know the location and use 'mini.files' to quickly navigate to it.

**Comment by assistcontrol** [comment]
> It is possible, as every source can have "current working directory". So with `builtin.files()` you can do `MiniPick.builtin.files(nil, { source = { cwd = '../foo/bar' } })`. If you plan to do that often interactively, you can [modify `files` in registry](https://github.com/echasnovski/mini.nvim/blob/0f93724569ebf27b6b3bc7130d8044a9020a8cc5/doc/mini-pick.txt#L1135-L1140) and use `:Pick files cwd='../foo/bar'`.

That first incantation is a little awkward, but works perfectly for me. Thanks!

> To be honest, I did not realize that picking something from different directory is a common thing. I'd make it more prominent in the documentation.

I’d encourage that! I (hope) I’m not the only person who’s wanted to do that, because if so then I’ve missed the point somewhere along the way.

> Maybe this is a workflow thing. I use [`setup_auto_root()`](https://github.com/echasnovski/mini.nvim/blob/0f93724569ebf27b6b3bc7130d8044a9020a8cc5/doc/mini-misc.txt#L119) and can't really remember when I needed to pick something from outside of the root. If I need something, then I usually know the location and use 'mini.files' to quickly navigate to it.

My use case there is primarily mini.starter. I have a couple projects that I often want to work on or examine a particular file in, and spawning nvim and browsing one of those dirs is a very common workflow for me.

**Comment by echasnovski** [comment]
> I saw this issue after I filed mine, huge apologies. TLDR is Pick grep searches .git folder, which isn't really useful. Excellent plugin!
> 
> #516

Duplicating from #516: suggested approach is to set up [custom `rg` config file](https://github.com/BurntSushi/ripgrep/blob/master/GUIDE.md#configuration-file). Thanks @Keyhart for the tip!

**Comment by echasnovski** [comment]
>     1. file first matching. For example, `foo/abc.ts` and `src/foo.ts`, after I input `foo`, I want `src/foo.ts` to be the first matched. Like `fzf --delimiter / --nth -1,..`

Yes, this is working as intended because matching algorithm is the same by default for any picker. If you add `.` or `t` (as in first letter of extension), it should put `src/foo.ts` first because its match has smaller width.

>     2. `:Pick grep` unable to paste to input?

What do you mean by "paste to input"? `builtin.grep` takes `pattern` as local option, which if not supplied asks user to input it interactively. You can do at least these:
- Execute `:Pick grep` and then paste manually to the interactive input (via `<C-r>` and register name, for example).
- Execute `:Pick grep pattern='<paste here manually>'`.
- Execute `:Pick grep pattern='<cword>'` to search for word under cursor.

**Comment by fannheyward** [comment]
> `:Pick grep` unable to paste to input?

Sorry, my typo.  I can't paste in `:Pick grep_live`.

**Comment by echasnovski** [comment]
> > `:Pick grep` unable to paste to input?
> 
> Sorry, my typo. I can't paste in `:Pick grep_live`.

What do you mean by "paste"? There is a [paste action](https://github.com/echasnovski/mini.nvim/blob/0f93724569ebf27b6b3bc7130d8044a9020a8cc5/doc/mini-pick.txt#L514) with default mapping of `<C-r>`. So you can `:Pick grep_live` followed by `<C-r>` and register name. If you want some automated functionality like preset query with certain word (like word under cursor), then yes, it needs extra manipulation along the lines of [this comment](https://github.com/echasnovski/mini.nvim/issues/513#issuecomment-1762785125).

**Comment by fannheyward** [comment]
> What do you mean by "paste"?

Copy something from other then paste to `grep_live` by system clipboard.

**Comment by echasnovski** [comment]
> > What do you mean by "paste"?
> 
> Copy something from other then paste to `grep_live` by system clipboard.

System clipboard can be accessed via one of `+` or `*` registers. Take a look at `:h quoteplus` and `:h quotestar`.

So on my machine I can select text from this page (select, click with right mouse button, select "Copy"). Then go into Neovim, execute `:Pick grep_live` followed by `<C-r>` and `+`. This will add system clipboard entry to the query.

**Comment by Jackevansevo** [comment]
I'm triggering mini.pick by the following mechanism

```lua
vim.keymap.set("n", "<leader>f",  ":Pick files<CR>", { noremap = true, silent = true })
vim.keymap.set("n", "<leader>/",  ":Pick grep<CR>", { noremap = true, silent = true })
vim.keymap.set("n", "<leader>b",  ":Pick buffers<CR>", { noremap = true, silent = true })
```

I'm finding neither `<C-n>` / `<Down>` or `<C-p>` / `<Up>` bindings are working for me (as suggested in the docs).

```
- Type special keys to perform |MiniPick-actions|. Here are some basic ones:
    - `<C-n>` / `<Down>` moves down; `<C-p>` / `<Up>` moves up.
```

**Comment by echasnovski** [comment]
Would you mind go into a bit more detail? Like:
- Do you press, for example, `<Leader>f` to start finding files. Does the floating window appear with the list of files?
- If window is active, what does pressing `<C-n>` or `<Down>` result in? Just nothing happens or happens something else (not moving the current item down)?
- Do other keys work? Like `<C-f>`, `<Tab>`, or `<CR>`?
- Just for a good measure: what terminal emulator and OS do you use?

**Comment by echasnovski** [comment]
> I'm on Mac OS & `iterm`, using `<leader>f` causes a floating window to appear, it's focussed and I can fuzzy search + open files in tabs or splits.
> 
> Most of the other bindings appear to be working, `<C-a>`, `<C-u>`, `<C-x>`, `<C-f>`, `<C-b>`
> 
> When pressing `<C-n>` / `<Down>` moves down; `<C-p>` / `<Up>` nothing is happening at all.

I think I have an idea what might be going on. What does `:hi MiniPickMatchCurrent` show? By default it is linked to `CursorLine` and if you have cleared it, then moving current item won't show any effect. However, there is still should be an effect. For example, right after you open the picker, press `<Down>` several times and then `<S-Tab>`. This would open the info view and `Current index` line should show number bigger than 1.

**Comment by Jackevansevo** [comment]
If I disable `mini.base16` (i.e. by not setting a colourscheme and re-opening vim) I can see that it's actually working as expected.

As soon as I do `colorscheme minicyan` or similar I can see it breaks. 

Running `:hi MiniPickMatchCurrent` consistently gives me:

```
MiniPickMatchCurrent xxx cterm= gui=links to CursorLine
```

whether I've got `mini.base16` enabled or not

**Comment by echasnovski** [comment]
> If I disable `mini.base16` (i.e. by not setting a colourscheme and re-opening vim) I can see that it's actually working as expected.
> 
> As soon as I do `colorscheme minicyan` or similar I can see it breaks.
> 
> Running `:hi MiniPickMatchCurrent` consistently gives me:
> 
> ```
> MiniPickMatchCurrent xxx cterm= gui=links to CursorLine
> ```
> 
> whether I've got `mini.base16` enabled or not

Yep, it is a 'mini.base16' issue. I forgot that floating window in it has the same background as cursor line. I'll fix it soon.

**Comment by oncomouse** [comment]
Is there any way `H.actions` could be made user-accessible? I'd like to make an action to mark and then move down, which should be as simple as:

~~~lua
require("mini.pick").setup({
  mappings = {
    mark = nil,
    mark_and_move = {
      char = "<C-X>",
      func = function(picker)
        H.actions.mark(picker)
        H.actions.move_down(picker)
      end
    }
  }
})

but isn't because of the way actions are defined, unless I'm missing something.

~~~

**Comment by oncomouse** [comment]
@echasnovski As a note, this works as well and can account for anything that would change the setup: `vim.api.nvim_input(vim.api.nvim_replace_termcodes(string.format("%s%s", MiniPick.config.mappings.mark, MiniPick.config.mappings.move_down), true, true, true))`

**Comment by echasnovski** [comment]
> @echasnovski As a note, this works as well and can account for anything that would change the setup: `vim.api.nvim_input(vim.api.nvim_replace_termcodes(string.format("%s%s", MiniPick.config.mappings.mark, MiniPick.config.mappings.move_down), true, true, true))`

Well, to be even more precise, it should account for possible picker-local mappings. Like this:
```lua
local mark_and_move_down = function()
  local mappings = MiniPick.get_picker_opts().mappings
  local keys = mappings.mark .. mappings.move_down
  vim.api.nvim_input(vim.api.nvim_replace_termcodes(keys, true, true, true))
end
require('mini.pick').setup({
  mappings = {
    mark_and_down = { char = '<C-d>', func = mark_and_move_down },
  },
})
```

**Comment by Jackevansevo** [comment]
> **Edit**: @Jackevansevo, I've pushed a 'mini.base16' update with tweaked highlight groups for 'mini.pick'. Should be all visible on latest `main`.

confirmed this is working perfectly now, cheers for the help 👍

**Comment by bethandtownes** [comment]
is there a way to enable preview automatically?

**Comment by echasnovski** [comment]
> is there a way to enable preview automatically?

Sorry, no. This doesn't play nicely with the possible asynchronous setting of items (i.e. there can be no items when picker is started meaning that preview is not properly defined). In theory, you can try to emulate it with `vim.api.nvim_input('\t')` after starting picker, but i wouldn't recommend.

**Comment by lostl1ght** [comment]
Hi, thanks for another great plugin!

There's a weird interaction with `mini.files`:
![image](https://github.com/echasnovski/mini.nvim/assets/61948252/7d1feaa3-7790-4a81-b1c7-267f33773f43)

`README.md` was opened with `:edit` and `pick.lua` was opened with `mini.files`. Current working directory is `~/.local/share/nvim/lazy/mini.pick` but it looks like `pick.lua` is from another directory (not a sub-directory of cwd). Is this intended and if so is there a way to change so that file names opened with `mini.files` would be displayed as relative to cwd?

This is how the same files are displayed in `telescope` and I'd like `mini.pick` to display the files the same way (apart from numbers and indicators):
![image](https://github.com/echasnovski/mini.nvim/assets/61948252/24232524-bcf6-4d19-836c-17a3e76bf25d)

**Comment by tani** [comment]
Can I use other commands instead of rg, fd and git? I want to use GNU find.
It looks like all the commands are hardcoded.

**Comment by echasnovski** [comment]
> Can I use other commands instead of rg, fd and git? I want to use GNU find. It looks like all the commands are hardcoded.

There is `builtin.cli()` for other CLI commands. For example:
- `:Pick cli command={'find','-type','f'}` using `:Pick` command.
- If you want a more similar experience to `builtin.files`, you can create custom `registry` entry. With something like this:
  ```
  MiniPick.registry.files_find = function()
    local show_with_icons =
      function(buf_id, items, query) return MiniPick.default_show(buf_id, items, query, { show_icons = true }) end
    local postprocess = function(lines)
      return vim.tbl_map(function(l) return l:gsub('^%./', '') end, lines)
    end
    return MiniPick.builtin.cli(
      { command = { 'find', '-type', 'f' }, postprocess = postprocess },
      { source = { show = show_with_icons } }
    )
  end
  ```

**Comment by echasnovski** [comment]
> Hi, thanks for another great plugin!
> 
> There's a weird interaction of `Pick buffers` with `mini.files`: ![image](https://user-images.githubusercontent.com/61948252/276765765-7d1feaa3-7790-4a81-b1c7-267f33773f43.png)
> 
> `README.md` was opened with `:edit` and `pick.lua` was opened with `mini.files`. Current working directory is `~/.local/share/nvim/lazy/mini.pick` but it looks like `pick.lua` is from another directory (not a sub-directory of cwd). Is this intended and if so is there a way to change so that file names opened with `mini.files` would be displayed as relative to cwd?
> 
> This is how the same files are displayed in `telescope` and I'd like `mini.pick` to display the files the same way (apart from numbers and indicators): ![image](https://user-images.githubusercontent.com/61948252/276765890-24232524-bcf6-4d19-836c-17a3e76bf25d.png)

@lostl1ght, this is relatively strange and definitely shouldn't be the case. I can not reproduce, though. The `builtin.buffers()` [use `:buffers`](https://github.com/echasnovski/mini.nvim/blob/e9a0716013abb8d2d6f8d77331600fb99eb15d51/lua/mini/pick.lua#L1344) under the hood which does show buffer names relative to current working directory.

What does `:buffers` show?

And just to be sure:
- Is 'README.md' also from '~/.local/share/nvim/lazy/mini.pick'? 
- What does `:pwd` show prior to reproducing this `:Pick buffers`?

**Comment by tani** [comment]
@echasnovski Thank you for the assistance. `Pick cli` is awesome!

**Comment by lostl1ght** [comment]
> Is 'README.md' also from '~/.local/share/nvim/lazy/mini.pick'?

Yes.

> What does `:pwd` show prior to reproducing this `:Pick buffers`?

`/home/<user>/.local/share/nvim/lazy/mini.pick`

> What does `:buffers` show?

This:
```
  2 %a   "README.md"                    line 29
  3      "~/.local/share/nvim/lazy/mini.pick/lua/mini/pick.lua" line 20
```

So from my understanding there's nothing really wrong with anything here. It's just that `mini.files` uses absolute paths to open files and so `:buffers` displays the absolute paths, am I right? But is there a way to display shortened paths?

**Comment by JSteitz** [comment]
It would be nice to have an option to change the default icons that are not coming from nvim-web-devicons.
In my use case, I add 1 extra space after the icon to get a nice spacing between icon and text.
This is probably easier to do than allowing to change how the lines are rendered.
That also allows us to set our favorite icon for Directory and Unknown files.

![image](https://github.com/echasnovski/mini.nvim/assets/677821/332edd8a-0a6e-4301-a15c-0950affb0ef1)

**Comment by echasnovski** [comment]
> It would be nice to have an option to change the default icons that are not coming from nvim-web-devicons.

I was waiting for someone to request it with a valid use case :)
Yeah, what you describe seems like one. I'll add is soon as `opts.icons` in `default_show()`.

**Comment by echasnovski** [comment]
> Huh, strange. I can describe the steps in more detail:
> 
>     1. `cd` into `~/.local/share/nvim/lazy/mini.pick`
> 
>     2. `nvim`
> 
>     3. `:edit README.md`
> 
>     4. `=MiniFiles.open()`
> 
>     5. Put the cursor on the `lua` folder
> 
>     6. Press `l` 3 times to open `pick.lua`
> 
>     7. `:buffers` or `Pick buffers` display buffers as I described

@lostl1ght, yes, very strange. I could finally reproduce with minimal config containing only 'mini.files'. I'll look into it.

**Comment by JSteitz** [comment]
Do we have the option to sort the buffer list by last used?

Use Case:
If you work on multiple files (more than 2) and have to switch between them a lot, then it is way more convenient to have them sorted by last used. This way, you do not have to think so hard about how the buffer is called because you already know that it is one of the top results, so you can just move a few steps down instead of filtering/searching.

Currently, my buffers look like this:
1: Main file I work on
2 - 10: Many files that I opened for reference
11: My second important file I work on
12: My third related file I work on

**Comment by echasnovski** [comment]
> Do we have the option to sort the buffer list by last used?
> 
> Use Case: If you work on multiple files (more than 2) and have to switch between them a lot, then it is way more convenient to have them sorted by last used. This way, you do not have to think so hard about how the buffer is called because you already know that it is one of the top results, so you can just move a few steps down instead of filtering/searching.
> 
> Currently, my buffers look like this: 1: Main file I work on 2 - 10: Many files that I opened for reference 11: My second important file I work on 12: My third related file I work on

No, not in the `builtin.buffers()`. Implementing that would need custom tracking instead of relying on built-in `:buffers`. The dedicated functionality of such tracking is currently planned in one of the next modules.

In the meantime, you can look into the following solutions:
- Use [`oldfile()`](https://github.com/echasnovski/mini.nvim/blob/b94d010c61ba46474af78536873db89492b251a4/doc/mini-bracketed.txt#L377) from 'mini.bracketed' (mapped to `[o` and `]o` by default).
- Write your own picker using `vim.fn.expand('#n')` functionality. Something like this:
  ```lua
  MiniPick.registry.buffers_recent = function()
    local items = {}
    for i = 1, 100 do
      local path = vim.fn.expand('#' .. i)
      if path ~= '' and vim.fn.filereadable(path) == 1 then table.insert(items, vim.fn.fnamemodify(path, ':~:.')) end
    end

    local show_with_icons =
      function(buf_id, items, query) MiniPick.default_show(buf_id, items, query, { show_icons = true }) end

    return MiniPick.start({ source = { items = items, name = 'Recent buffers', show = show_with_icons } })
  end
  ```
  It is not ideal (includes visited help files), but concise. 
  Then use it with `:Pick buffers_recent`.

**Comment by echasnovski** [comment]
@lostl1ght, the absolute path in `:buffers` after opening in 'mini.files' should be fixed on latest `main`. The issue indeed was that 'mini.files' used absolute path with `:edit` command. Although this is a more robust approach, it indeed has the issue that you found. I thought it shouldn't matter because `nvim_buf_get_name()` always returns full path and `:buffers` can handle that.

Apparently, it can handle that, but only after another `cd` (even if in the same directory). So everything is OK after these set of commands:
```
:cd ~/.local/share/nvim/lazy/mini.pick
:edit ~/.local/share/nvim/lazy/mini.pick/lua/mini/pick.lua
:cd .
```

The reason why I couldn't reproduce it is because I have `setup_auto_root()` from 'mini.misc' which does equivalent of `cd` on every `BufEnter`.

**Comment by lostl1ght** [comment]
Works great. Thank you!

**Comment by lostl1ght** [comment]
Also, I've finally figured out how to reproduce a bug(?) with the cursor and `cmdheight = 0`:
![image](https://github.com/echasnovski/mini.nvim/assets/61948252/40b6add0-4e64-4362-bb8e-883e662bd34a)
Can you see how the cursor is on top of the floating window?
This is very annoying if one uses folke's `noice.nvim`.
If `cmdheight > 0` then the cursor just stays in the cmdline.

**Comment by echasnovski** [comment]
> Also, I've finally figured out how to reproduce a bug(?) with the cursor and `cmdheight = 0`:

Sigh... Yeah, this is expected and seems to be the issue with Neovim itself, as I simply couldn't find a way to not show cursor reliably. Hence a "hack" to place it in command line.

I'll ask Neovim core about this.

**Comment by lostl1ght** [comment]
Is there a reason why `mini.pick` uses what it seems to be a unicode character as a cursor instead of an actual cursor?

**Comment by echasnovski** [comment]
> Is there a reason why `mini.pick` uses what it seems to be a unicode character as a cursor instead of an actual cursor?

Do you mean as caret in prompt? That is because it is not a cursor but a visual representation of a caret in a prompt. As prompt is displayed in floating window's title, there is no way to actually place a cursor in Insert mode there.

**Comment by lostl1ght** [comment]
> Do you mean as caret in prompt?

Yes.

> there is no way to actually place a cursor in Insert mode there.

I see. Thank you.

**Comment by oncomouse** [comment]
> Do we have the option to sort the buffer list by last used?

@JSteitz I think there might still be some weird edge cases lingering, but I [ported telescope's `sort_lastused` option into MiniPick in my dotfiles](https://github.com/oncomouse/dotfiles/blob/37de1293d98ed4cc5162e7bdd182ddef91cc7955/conf/vim/lua/dotfiles/plugins/mini-nvim.lua#L448-L499).

You can use `:Pick buffers sort_lastused=true` to activate it.

**Comment by echasnovski** [comment]
> > Do we have the option to sort the buffer list by last used?
> 
> @JSteitz I think there might still be some weird edge cases lingering, but I [ported telescope's `sort_lastused` option into MiniPick in my dotfiles](https://github.com/oncomouse/dotfiles/blob/37de1293d98ed4cc5162e7bdd182ddef91cc7955/conf/vim/lua/dotfiles/plugins/mini-nvim.lua#L448-L499).
> 
> You can use `:Pick buffers sort_lastused=true` to activate it.

I remember looking at `lastused` from `getbufinfo()` and it didn't *quite* do what is intuitively expected. It is really clear what "last used" actually means and it tracks timestamp up to seconds (so quick buffer change with `]b` or `]o` from 'mini.bracketed' doesn't produce good result). That's mostly the reasons why it is not included in `builtin.buffers()`.

**Comment by oskarkook** [comment]
I quickly wrote a picker that replicates fzf-lua's `lsp_document_symbols()` behaviour (see below).

> Are configuration and function/picker names intuitive enough?

Yes

> Are default values of settings convenient for you?

Yes

> Is documentation and examples clear enough?

Yes

> Does default matching algorithm give relevant enough results?

Yes, although probably I'd customize it for this picker a bit. But seems easy enough to customize it judging by the docs.

```lua
MiniPick.registry.lsp_document_symbols = function()
  local bufnr = vim.api.nvim_get_current_buf()

  vim.lsp.buf_request_all(
    bufnr,
    'textDocument/documentSymbol',
    { textDocument = vim.lsp.util.make_text_document_params(bufnr) },
    function(lsp_responses)
      local items = {}

      local function process_result(result, depth)
        table.insert(items, {
          text = string.rep('  ', depth) .. '[' .. vim.lsp.protocol.SymbolKind[result.kind] .. '] ' .. result.name,
          bufnr = bufnr,
          lnum = result.range['start'].line + 1,
          end_lnum = result.range['end'].line + 1,
          col = result.range['start'].character + 1,
          end_col = result.range['end'].character + 1,
        })
        if result['children'] ~= nil then
          for _, child in ipairs(result.children) do
            process_result(child, depth + 1)
          end
        end
      end

      for _, lsp_response in ipairs(lsp_responses) do
        for _, result in ipairs(lsp_response.result) do
          process_result(result, 0)
        end
      end

      MiniPick.start({
        source = {
          items = items,
          name = 'LSP document symbols'
        }
      })
    end
  )
end
```

**Comment by echasnovski** [comment]
Thanks for the feedback, @oskarkook!

> I quickly wrote a picker that replicates fzf-lua's lsp_document_symbols() behaviour (see below).

This looks sooo nice! I have a more general implementation based on `vim.lsp.buf` module, but yours looks much better.

**Comment by oskarkook** [comment]
Is there a reason why `lnum`, `col` etc are 1-based? It seems I need to always add 1 to values that are received from vim APIs.

```lua
MiniPick.registry.document_diagnostics = function()
  local items = {}
  local results = vim.diagnostic.get()

  for _, result in ipairs(results) do
    table.insert(items, {
      text = result.message,
      bufnr = result.bufnr,
      lnum = result.lnum + 1,
      end_lnum = result.end_lnum + 1,
      col = result.col + 1,
      end_col = result.end_col + 1,
    })
  end

  MiniPick.start({
    source = {
      items = items,
      name = 'Document diagnostics'
    }
  })
end
```

**Comment by echasnovski** [comment]
> Is there a reason why `lnum`, `col` etc are 1-based? It seems I need to always add 1 to values that are received from vim APIs.
> ...
> Mainly I'm thinking whether the plugin's API should match vim's? But it's a mild enough difference, fine as it is as well.

That is the thing: 1-based numeration **is** what Vim uses. It is Neovim's API what uses 0-based positions (and sometimes mixed, like in `nvim_win_{get,set}_cursor()`). See `:h getqflist()`, for example.

Besides, Lua also uses 1-based indexing. Both of those reasons are why 'mini.nvim' early on tried to use 1-based indexes.

**Comment by oskarkook** [comment]
Roger :). The plugin works great for my use cases! Adding extra pickers was easy and overall I'm super happy with it. I can even empirically perceive a lower latency compared to alternatives. Thanks a lot for the work you do on mini.nvim!

**Comment by JSteitz** [comment]
> > Do we have the option to sort the buffer list by last used?
> 
> @JSteitz I think there might still be some weird edge cases lingering, but I [ported telescope's `sort_lastused` option into MiniPick in my dotfiles](https://github.com/oncomouse/dotfiles/blob/37de1293d98ed4cc5162e7bdd182ddef91cc7955/conf/vim/lua/dotfiles/plugins/mini-nvim.lua#L448-L499).
> 
> You can use `:Pick buffers sort_lastused=true` to activate it.

@oncomouse 
Thanks, that would have been my next attempt.
After testing it, I can see that some buffers are not sorted correctly, especially buffers for files from different directories.

So I tried to do it myself and this is the result:

```lua
MiniPick.registry.buffers = function(local_opts, opts)
  local_opts = vim.tbl_deep_extend('force', { sort_lastused = false, include_current = true, include_unlisted = false }, local_opts or {})

  local cur_buf_id, include_current = vim.api.nvim_get_current_buf(), local_opts.include_current
  local flag_unlisted = local_opts.include_unlisted and "!" or ""
  local flag_recent = local_opts.sort_lastused and "t" or ""
  local command = string.format("buffers%s %s", flag_unlisted, flag_recent)
  local buffers_output = vim.api.nvim_exec2(command, { output = true }).output
  local items = {}

  for _, l in ipairs(vim.split(buffers_output, '\n')) do
    local buf_str, name = l:match('^%s*%d+'), l:match('"(.*)"')
    local buf_id = tonumber(buf_str)
    local item = { text = name, bufnr = buf_id }
    if buf_id ~= cur_buf_id or include_current then table.insert(items, item) end
  end

  local show = MiniPick.config.source.show or function(buf_id, items, query)
    MiniPick.default_show(buf_id, items, query, { show_icons = true })
  end

  return MiniPick.start(vim.tbl_deep_extend('force', { source = { name = 'Buffers', show = show } }, opts or {}, { source = { items = items } }))
end
```

`:buffers` has a flag to get all buffers already sorted by last used. So I'm just using that. Everything else is the same, as the `builtin.buffers`. @echasnovski is this solution good and simple enough to include?

@echasnovski
`mini.pick` is in beta, but the documentation reads like it is already included in `mini.nvim`.

If that is so, then I get the error that it does not find `mini.pick`.
Otherwise, it is misleading in the documentation and would be better if there was a note next to the module "(beta) not released" in the README.

Can someone give an example of how I can filter out directories? Don't know much about lua and neovim api :/
I am not convinced that this is enough to be a feature, more a preference in my use case.

**Comment by JSteitz** [comment]
@echasnovski 

I have sometimes performance issues with `builtin.grep_live` where my CPU maxes out.
The project itself is big (many node_modules and other dependencies).
Does live_grep use the `.gitignore`?

This happens mostly initially after first startup of neovim. I never had this issue with Telescope using ripgrep.
How can I help with that?

**Comment by echasnovski** [comment]
> `:buffers` has a flag to get all buffers already sorted by last used. So I'm just using that. Everything else is the same, as the `builtin.buffers`. @echasnovski is this solution good and simple enough to include?

Hmm... This is indeed an interesting suggestion which I wasn't actively aware of. I'll have a look into how it works on practice.

> @echasnovski `mini.pick` is in beta, but the documentation reads like it is already included in `mini.nvim`.
> 
> If that is so, then I get the error that it does not find `mini.pick`. Otherwise, it is misleading in the documentation and would be better if there was a note next to the module "(beta) not released" in the README.

Yes, it is in public beta.
Yes, it is included in a 'mini.nvim'.
No it is not misleading.

When exactly does the error occur? If you use 'mini.nvim', make sure that you don't use `stable` branch or tag version.

> Can someone give an example of how I can filter out directories? Don't know much about lua and neovim api :/
> I am not convinced that this is enough to be a feature, more a preference in my use case.

From where do you want to filter out directories?

**Comment by echasnovski** [comment]
I've just pushed several updates.

------

The most [visible one](https://github.com/echasnovski/mini.nvim/commit/f39a455770aefdecda114370784c7dd7f9f7bf74) is that all CLI tools are now called with a minimal set of flags needed for picker to work properly. In practice, that means that both `fd`/`rg` in `builtin.files()` and `rg` in `builtin.grep()`/`builtin.grep_live()` don't recognize hidden files by default. To customize that, [follow help](https://github.com/echasnovski/mini.nvim/blob/7af4878bd0b19fcef385e45715f418165eff1218/doc/mini-pick.txt#L77-L82).

------

@lostl1ght, 'mini.pick' now work for plain `cmdheight=0`, but not with 'folke/noice.nvim' (unfortunately). We've chatted with Folke a bit (as this is a bit of a complicated topic spanning 'mini.pick', Noice, and Neovim core), and the best approach here seems to let each plugin do its thing:
- 'mini.pick' places cursor in command line to avoid it being visible in picker window.
- 'noice.nvim' now [practices custom dark magic ways](https://github.com/folke/noice.nvim/commit/0cca940561d4b723cb52ba9e4ec239fceb36f146) of making making sure that cursor is up to date. Unfortunately, together with absent command line (as Noice itself is a command line), this means that cursor will always be shown in picker window.
- But the best outcome here would be for Neovim core to implement proper ways of hiding and showing cursor, as it is not possible now. This needs some work, dedication, and C knowledge, so probably won't be done any time soon.

------

Another update is for you, @JSteitz: `default_show()` now [supports custom icons](https://github.com/echasnovski/mini.nvim/blob/7af4878bd0b19fcef385e45715f418165eff1218/doc/mini-pick.txt#L940-L943). To use them, update every picker with which you want to use custom icons (like `files`) to use `show_with_icons()` similar to [this comment](https://github.com/echasnovski/mini.nvim/issues/513#issuecomment-1772239061). I know, it is not super scalable, but that is the price of having granular control over pickers.

**Comment by echasnovski** [comment]
> `:buffers` has a flag to get all buffers already sorted by last used. So I'm just using that. Everything else is the same, as the `builtin.buffers`. @echasnovski is this solution good and simple enough to include?

@JSteitz, I've looked into this and `:buffers t` indeed doesn't always give an up to date results. I think this is mostly because "last used" is not really well defined anywhere. For example, ibhagwan/fzf-lua also [ended up making custom solution](https://github.com/ibhagwan/fzf-lua/blob/04e7382528f3f4970cb1cd559d06ae6f7b44560e/lua/fzf-lua/providers/buffers.lua#L88-L108).

As there is a planned alternative module with a similar functionality (but for files in general), I think `sort_lastused` for buffer won't be part of `builtin.buffers()`.

**Comment by JSteitz** [comment]
> Another update is for you, @JSteitz: default_show() now [supports custom icons](https://github.com/echasnovski/mini.nvim/blob/7af4878bd0b19fcef385e45715f418165eff1218/doc/mini-pick.txt#L940-L943). To use them, update every picker with which you want to use custom icons (like files) to use show_with_icons() similar to https://github.com/echasnovski/mini.nvim/issues/513#issuecomment-1772239061. I know, it is not super scalable, but that is the price of having granular control over pickers.

Not a problem, have to create keybinds to call the methods anyway, so I'm good with that :+1: 

> As there is a planned alternative module with a similar functionality (but for files in general), I think sort_lastused for buffer won't be part of builtin.buffers().

I think that is a reasonable decision.
If people want to have such a feature now, they could just use my code snippet meanwhile.

**Comment by fightingdreamer** [comment]
I really like this plugin, feels very correct — I'm curious if something like `require('mini.pick').load_extension()` can be implemented to allow extra sources be implemented as plugins?

**Comment by echasnovski** [comment]
> I'm curious if something like `require('mini.pick').load_extension()` can be implemented to allow extra sources be implemented as plugins?

The suggested way to do this is with [`MiniPick.registry`](https://github.com/echasnovski/mini.nvim/blob/f699158e6a654d2778e446d98de2881d7ab1f5fb/doc/mini-pick.txt#L1142). This is a more flexible approach at the cost of being slightly more explicit. It can be utilized in the following ways:
- Have "extension" explicitly add picker with `require('min.pick').registry.my_ext = function(local_opts) ... end`. This works but has the downside of having 'mini.pick' being loaded strictly prior to that extension.
- Have user make that assignment themselves (in the config) instead of `load_extenstion()`. This requires "extension" to export either full picker itself or a basic functionality to get items and let the user make the call to `MiniPick.start()`.

**Comment by jmbaur** [comment]
> > > What do you mean by "paste"?
> > 
> > 
> > Copy something from other then paste to `grep_live` by system clipboard.
> 
> System clipboard can be accessed via one of `+` or `*` registers. Take a look at `:h quoteplus` and `:h quotestar`.
> 
> So on my machine I can select text from this page (select, click with right mouse button, select "Copy"). Then go into Neovim, execute `:Pick grep_live` followed by `<C-r>` and `+`. This will add system clipboard entry to the query.

I am having issues using this solution using MiniPick on a remote machine over SSH. Any proposed solution there?

**Comment by echasnovski** [comment]
> I am having issues using this solution using MiniPick on a remote machine over SSH. Any proposed solution there?

Does clipboard work in regular scenarios? Like typing `"+p` (or `"*p`) in inside modified text buffer leads to pasting text from the clipboard? As history shows, system clipboard and SSH is not a great pairing and can be a trouble to set up. I don't really have experience here, unfortunately.

**Comment by jmbaur** [comment]
> Does clipboard work in regular scenarios? Like typing `"+p` (or `"*p`) in inside modified text buffer leads to pasting text from the clipboard? As history shows, system clipboard and SSH is not a great pairing and can be a trouble to set up. I don't really have experience here, unfortunately.

Yes, when nvim is running on the same machine the terminal is running on, pasting regularly works fine. Using mini-pick with `<C-r>` + register name also works great in this scenario.

**Comment by echasnovski** [comment]
> Yes, when nvim is running on the same machine the terminal is running on, pasting regularly works fine. Using mini-pick with `<C-r>` + register name also works great in this scenario.

My question was a bit ambiguous. What I meant by "work in regular scenarios" is "does regular pasting inside SSH work, i.e. like typing ...".

What 'mini.pick' does with `<C-r>` is [quite literally using the content of the register](https://github.com/echasnovski/mini.nvim/blob/f699158e6a654d2778e446d98de2881d7ab1f5fb/lua/mini/pick.lua#L2425-L2426). If regular pasting with system clipboard doesn't work for some reason, then `<C-r>` in 'mini.pick' won't work. And vice versa: if one works, then should the other.

**Comment by jmbaur** [comment]
> My question was a bit ambiguous. What I meant by "work in regular scenarios" is "does regular pasting inside SSH work, i.e. like typing ...".
> 
> What 'mini.pick' does with `<C-r>` is [quite literally using the content of the register](https://github.com/echasnovski/mini.nvim/blob/f699158e6a654d2778e446d98de2881d7ab1f5fb/lua/mini/pick.lua#L2425-L2426). If regular pasting with system clipboard doesn't work for some reason, then `<C-r>` in 'mini.pick' won't work. And vice versa: if one works, then should the other.

Ah, then yes this does not work over a remote session. I will look into getting OSC52 paste working in nvim so the system clipboard can be used!

**Comment by hbiel** [comment]
Thanks for another great plugin! I already set it up to replace telescope and am using it's vim.ui.select() implementation.

I recently tried https://github.com/biozz/whop.nvim though and experienced a minor issue.
The plugin let's you choose from a set of (mostly shell) commands and executes it against the current buffer.

Using mini.pick the command however seems to be executed on the mini.pick buffer instead.
Wrapping the on_choice function with vim.schedule_wrap() fixes it.

I feel like this would be good to handle within mini.pick as it might potentially break other plugins as well.

**Comment by echasnovski** [comment]
> Thanks for another great plugin! I already set it up to replace telescope and am using it's vim.ui.select() implementation.
> 
> I recently tried https://github.com/biozz/whop.nvim though and experienced a minor issue. The plugin let's you choose from a set of (mostly shell) commands and executes it against the current buffer.
> 
> Using mini.pick the command however seems to be executed on the mini.pick buffer instead. Wrapping the on_choice function with vim.schedule_wrap() fixes it.
> 
> I feel like this would be good to handle within mini.pick as it might potentially break other plugins as well.

That's a good use case, thanks!

Using `vim.schedule_wrap()` for `on_choice()` is OK, but might not work with implementation. I'll look into this.

**Comment by bray** [comment]
Hi - the README says "Out of the box support for 'ignorecase' and 'smartcase'."

I have both options set, but picking (e.g. `MiniPick.builtin.files()`) is case-sensitive.  With the same config, both Telescope and fzf.vim are smartcase.  I tried a minimal config with only mini.pick enabled, but no luck.

It's using ripgrep, so I wonder if the fact that rg defaults to case-sensitive is the issue?  If so, is there a way to pass in flags to the rg command?

Or am I missing something more obvious?  Thanks!

```
❯ nvim --version
NVIM v0.10.0-dev-aab06ed
Build type: RelWithDebInfo
LuaJIT 2.1.1694316387
```

**Comment by echasnovski** [comment]
> Hi - the README says "Out of the box support for 'ignorecase' and 'smartcase'."
> 
> I have both options set, but picking (e.g. `MiniPick.builtin.files()`) is case-sensitive. With the same config, both Telescope and fzf.vim are smartcase. I tried a minimal config with only mini.pick enabled, but no luck.
> 
> It's using ripgrep, so I wonder if the fact that rg defaults to case-sensitive is the issue? If so, is there a way to pass in flags to the rg command?
> 
> Or am I missing something more obvious? Thanks!
> 
> ```
> ❯ nvim --version
> NVIM v0.10.0-dev-aab06ed
> Build type: RelWithDebInfo
> LuaJIT 2.1.1694316387
> ```

It does respect those options for pickers which use `default_match()` (that is, not `grep_live()`). So I'd guess it is one of the two:
- Options `'ignorecase'` and `'smartcase'` are not set. What does `:=vim.o.ignorecase` and `:=vim.o.smartcase` show? They both should be `true` for most pickers to query with smart case.
- You used `builtin.grep_live` without `rg` being configured to use `--smart-case`.

(I guess that line in README could benefit from adding "... for 'ignorecase' and 'smartcase' *options*")

**Comment by bray** [comment]
Thanks for the quick reply.  I mentioned that I do have both options 'ignorecase' and 'smartcase' set.  `:=vim.o.ignorecase` and `:=vim.o.smartcase` both show `true`.  And I'm not using grep_live, only `MiniPick.builtin.files()`.

I also tried `:= MiniPick.builtin.files({ tool = 'rg' })` and 'fd' and 'git' (and 'fallback' but that strangely shows no results).  All of them are case-sensitive.

**Comment by echasnovski** [comment]
> Thanks for the quick reply. I mentioned that I do have both options 'ignorecase' and 'smartcase' set. `:=vim.o.ignorecase` and `:=vim.o.smartcase` both show `true`. And I'm not using grep_live, only `MiniPick.builtin.files()`.
> 
> I also tried `:= MiniPick.builtin.files({ tool = 'rg' })` and 'fd' and 'git' (and 'fallback' but that strangely shows no results). All of them are case-sensitive.

Sorry to hear that, but I can not reproduce. When I execute `:= MiniPick.builtin.files({ tool = 'rg' })`, matching is most certainly smart-case. For example:
- For query "li" it matches both "LICENSE" and "mini-align.md".
- For query "LI" it matches only 'LICENSE".
- For query 'lI" it matches none of them.

Besides those two cases I can not think about any other reasons why this should happen. I am pretty sure that it works as intended because also all tests for this behavior passed very recently.

>  I tried a minimal config with only mini.pick enabled, but no luck.

How did you do it?

**Comment by bray** [comment]
Oh - I trimmed down my config to only run this plugin and ran nvim as normal.  But I tracked it down to an option I have set elsewhere: `vim.opt.regexpengine = 1`.  If I remove this (and have it automatically select the regexp engine), that fixes smartcase here.

I have that set to 1 for performance reasons (not related to this plugin).  I'm open to removing it for now to see if I'm still seeing performance issues elsewhere.  But do you know why the old regexp engine breaks smartcase here?  Is there a workaround we can add to our config if we use the old engine?

And either way, maybe worth a mention to this in the relevant help page?

**Comment by echasnovski** [comment]
> Oh - I trimmed down my config to only run this plugin and ran nvim as normal. But I tracked it down to an option I have set elsewhere: `vim.opt.regexpengine = 1`. If I remove this (and have it automatically select the regexp engine), that fixes smartcase here.
> 
> I have that set to 1 for performance reasons (not related to this plugin). I'm open to removing it for now to see if I'm still seeing performance issues elsewhere. But do you know why the old regexp engine breaks smartcase here? Is there a workaround we can add to our config if we use the old engine?
> 
> And either way, maybe worth a mention to this in the relevant help page?

Interesting. That is maybe the third time I've been aware that this option exists, so didn't look at it. With it, I can reproduce. Thanks for the find!

I'll take a look why this is happening and will try to fix it.

**Comment by bray** [comment]
One more thing: with 'smartcase', when using `/`, you can change the behavior to force case-sensitive by adding "\C" anywhere in the pattern (see `:h /\C`)?  e.g. if I want "foo" to not match "Foo", I can type "foo\C" and it will no longer match "Foo" or "Fbaroo", but still be a fuzzy match, so it'll match "foo" and "fbaroo".

Is there a way to do that in this picker?  That would be super useful!

**Comment by echasnovski** [comment]
There are several new updates here.

------

@hbiel, `ui_select()` should now call `on_choice` when target window (and buffer) is current. It indeed seems to be the better behavior for `ui_select`.

------

@bray, `regexpengine=1` should now not block ignorecase/smartcase matching. The reason was that Vimscript's `match()` doesn't work well with `[[:upper:]]` regex atom when `regexpengine` is 1. And it was used to check if query is smart case.

------

> One more thing: with 'smartcase', when using `/`, you can change the behavior to force case-sensitive by adding "\C" anywhere in the pattern (see `:h /\C`)? e.g. if I want "foo" to not match "Foo", I can type "foo\C" and it will no longer match "Foo" or "Fbaroo", but still be a fuzzy match, so it'll match "foo" and "fbaroo".
> 
> Is there a way to do that in this picker? That would be super useful!

No, I don't think this will be a part of `default_match()`. It already has quite complicated logic about how it parses query.

**Comment by hbiel** [comment]
> `ui_select()` should now call `on_choice` when target window (and buffer) is current. It indeed seems to be the better behavior for `ui_select`.

Works like a charm! I also didn't know about `vim.api.nvim_win_call()`, which seems to be a better fit in this case.
Thanks!

**Comment by JulesNP** [comment]
Using the open in split or new tab actions for `:Pick help` don't seem to work as expected, it's creating a new split of the current buffer first.

**Comment by echasnovski** [comment]
> Using the open in split or new tab actions for `:Pick help` don't seem to work as expected, it's creating a new split of the current buffer first.

It works as documented: it [first opens split and then chooses](https://github.com/echasnovski/mini.nvim/blob/e8a413b1a29f05bb556a804ebee990eb54479586/doc/mini-pick.txt#L472-L475). Otherwise, to be consistent, plain choosing with `<CR>` should open help in the current window and not in split. I understand that this is not ideal, but I've tried making it work and didn't find any hackery-free solution. May look again in the future.

**Comment by echasnovski** [comment]
@JulesNP, today I realized that I can make `builtin.help` help custom mappings for `<C-s>`, `<C-v>`, and `<C-t>`. So those should now open the help window in horizontal, vertical, and tabpage splits.

**Comment by marcusandre** [comment]
> @JulesNP, today I realized that I can make `builtin.help` help custom mappings for `<C-s>`, `<C-v>`, and `<C-t>`. So those should now open the help window in horizontal, vertical, and tabpage splits.

Awesome. Just updated and can report that it is working fine 👍

**Comment by JulesNP** [comment]
> @JulesNP, today I realized that I can make `builtin.help` help custom mappings for `<C-s>`, `<C-v>`, and `<C-t>`. So those should now open the help window in horizontal, vertical, and tabpage splits.

Thanks! Everything's working great with the latest changes. I've fully switched over from Telescope now in my config.

**Comment by bassamsdata** [comment]
I want to express my gratitude first for creating all these fantastic modules.

I've been utilizing `mini.Pick` since its inception, and it has proven to be excellent, especially in terms of its responsiveness and speed. It has nearly replaced my telescope workflow, particularly after the launch of `mini.extra`, for which I extend my congratulations.

While `mini.Pick` and `mini.extra` cover most of my needs, I still rely on telescope for one specific tool: the [telescope-undo](https://github.com/debugloop/telescope-undo.nvim) extension. This extension has been truly amazing, especially with its yanking capabilities for addition or deletion. My suggestion or rather a request is to consider creating a dedicated pick for undo/history functionality (or dedicated Module eg. `Mini.Unfasten` 🫣). This module could stand out by integrating git history seamlessly into the same timeline as undo 🔥, a feature that, as of now, is absent in the vim universe but present in VS Code.

**Additionally,** I'm curious if there's a way to customize the appearance of each pick. For instance, I'd like pick buffers or files to be much smaller and positioned in the bottom right, unlike grep_live, which I prefer to have in a larger window.

**On a related note,** I'm wondering if there's a method to implement grep_live specifically for help files (not the one for only help tags). I attempted to do so, but the outcome was less than satisfactory. Apologies for the less-than-ideal implementation.

<details>
<summary>My attempt for Pick help_grep</summary>

> Note: some of the code below borrowed and modified from another tele. ext.

```lua
    config = function(_, opts)
      local MiniPick = require("mini.pick")
      MiniPick.setup({})
      MiniPick.registry.help_files = function(local_opts)
        -- Parse options
        local_opts = vim.tbl_deep_extend("force", { prompt = "" }, local_opts or {})
        local prompt = local_opts.prompt
        local_opts.prompt = nil

        -- Get the paths of the help files
        local paths = vim.api.nvim_get_option("runtimepath")
        paths = vim.split(paths, ",")
        local dirs = {}
        for _, path in ipairs(paths) do
          local doc_path = path .. "/doc"
          if vim.fn.isdirectory(doc_path) == 1 then
            table.insert(dirs, doc_path)
          end
        end

        -- Start the picker based on CLI output
        MiniPick.builtin.cli({
          command = {
            "rg",
            prompt,
            "--color",
            "never",
            "--no-messages",
            "--line-buffered",
            "-g",
            "!*.log",
            "-n",
            "-H",
            unpack(dirs),
          },
          prompt_title = "Help Grep",
        }, local_opts)
      end
    end,
```


</details>


thank you for your time and assistance!

**Comment by echasnovski** [comment]
> I want to express my gratitude first for creating all these fantastic modules.
> 
> I've been utilizing `mini.Pick` since its inception, and it has proven to be excellent, especially in terms of its responsiveness and speed. It has nearly replaced my telescope workflow, particularly after the launch of `mini.extra`, for which I extend my congratulations.

Thanks for kind words! Much appreciated :pray: 

> While `mini.Pick` and `mini.extra` cover most of my needs, I still rely on telescope for one specific tool: the [telescope-undo](https://github.com/debugloop/telescope-undo.nvim) extension. This extension has been truly amazing, especially with its yanking capabilities for addition or deletion. My suggestion or rather a request is to consider creating a dedicated pick for undo/history functionality (or dedicated Module eg. `Mini.Unfasten` 🫣). This module could stand out by integrating git history seamlessly into the same timeline as undo 🔥, a feature that, as of now, is absent in the vim universe but present in VS Code.

Thanks for the suggestion! I'll think about it, but initial reaction is that this is unlikely to happen. There is a certain novel undo integration in 'mini.bracketed' (see [its `undo()` target](https://github.com/echasnovski/mini.nvim/blob/e1f49ae456f4e6e8b79e657965f295886fbe2101/doc/mini-bracketed.txt#L443)). I vividly remember having quite a few problems dealing with the undo history (as in, it can be done but there are many hidden pitfalls).

> **Additionally,** I'm curious if there's a way to customize the appearance of each pick. For instance, I'd like pick buffers or files to be much smaller and positioned in the bottom right, unlike grep_live, which I prefer to have in a larger window.

Yes, it is possible. All pickers in `MiniPick.builtin` and `MiniExtra.pickers` take two arguments: `local_opts` (specific to picker) and `opts` (configuration for `MiniPick.start()`).

Here is a quick example of how to create a new picker `help_right` inside a [`MiniPick.registry`](https://github.com/echasnovski/mini.nvim/blob/e1f49ae456f4e6e8b79e657965f295886fbe2101/doc/mini-pick.txt#L1153):

```lua
MiniPick.registry.help_right = function()
  local win_config =
    { relative = 'editor', anchor = 'SE', row = vim.o.lines, col = vim.o.columns, height = 20, width = 40 }
  return MiniPick.builtin.help({}, { window = { config = win_config } })
end
```

Having this executed (after `require('mini.pick').setup()`), you can now do `:Pick help_right`. Customize specific details of window config to your liking (see `:h nvim_open_win`).

> **On a related note,** I'm wondering if there's a method to implement grep_live specifically for help files (not the one for only help tags). I attempted to do so, but the outcome was less than satisfactory. Apologies for the less-than-ideal implementation.
> My attempt for Pick help_grep
> 
> > Note: some of the code below borrowed and modified from another tele. ext.
> 
> ```lua
>     config = function(_, opts)
>       local MiniPick = require("mini.pick")
>       MiniPick.setup({})
>       MiniPick.registry.help_files = function(local_opts)
>         -- Parse options
>         local_opts = vim.tbl_deep_extend("force", { prompt = "" }, local_opts or {})
>         local prompt = local_opts.prompt
>         local_opts.prompt = nil
> 
>         -- Get the paths of the help files
>         local paths = vim.api.nvim_get_option("runtimepath")
>         paths = vim.split(paths, ",")
>         local dirs = {}
>         for _, path in ipairs(paths) do
>           local doc_path = path .. "/doc"
>           if vim.fn.isdirectory(doc_path) == 1 then
>             table.insert(dirs, doc_path)
>           end
>         end
> 
>         -- Start the picker based on CLI output
>         MiniPick.builtin.cli({
>           command = {
>             "rg",
>             prompt,
>             "--color",
>             "never",
>             "--no-messages",
>             "--line-buffered",
>             "-g",
>             "!*.log",
>             "-n",
>             "-H",
>             unpack(dirs),
>           },
>           prompt_title = "Help Grep",
>         }, local_opts)
>       end
>     end,
> ```
> 
> thank you for your time and assistance!

I would advise not to try and implement specifically a `grep_live` version, as it it is quite involved if you want it to work smoothly (or at all even, really). If you must, I'd start from [`grep_live` source](https://github.com/echasnovski/mini.nvim/blob/e1f49ae456f4e6e8b79e657965f295886fbe2101/lua/mini/pick.lua#L1287) (there are couple of unexported helpers used, but they should be replaceable).

What you have is a variant of `builtin.grep`, which is a more reasonable idea. The general approach of using `rg` inside a dedicated runtime directories seems doable. The details of which arguments to use I'll leave to the audience :) One observation, though: `prompt_title` is not a part of [`builtin.cli` arguments](https://github.com/echasnovski/mini.nvim/blob/e1f49ae456f4e6e8b79e657965f295886fbe2101/doc/mini-pick.txt#L1139); you'd probably want a `{ source = { name = 'Help Grep' } }` as the second, `opts` argument.

That said, I'd suggest to go another route completely. Take a look at `:h helpgrep` or `:h lhelpgrep`. With it you can do something like this:
- `:helpgrep <pattern>` (like `:helpgrep window`). This creates a quickfix list (with `lhelpgrep` it creates a location list).
- Use `:Pick list scope='quickfix'` to navigate through quickfix list (or use location list if needed).

You can wrap those two steps in a single function if you'd like.

**Comment by Aphosis** [comment]
Hi ! Thanks a lot for `mini.pick`, it was a breeze to configure and works great ! Performance is really good too, which was my main reason for switching.

I've been wondering if there was something I could do to switch my `rg` configuration between pickers ?

For example, I'd like the default behavior of `grep_live` to use my default `rg` configuration, ignoring hidden and ignored files. But I'd also like to be able to call `grep_live` with another `rg` configuration, in case I need to find something in a hidden or ignored file.

To keep the configuration method consistent with the chosen solution, in the case of `rg` this would mean the ability, from `mini.pick`, to override `RIPGREP_CONFIG_PATH` for a given picker.

I saw that there is already support for [cwd](https://github.com/echasnovski/mini.nvim/blob/f2b89efbbad1943657e43f474fe308fceb63597e/lua/mini/pick.lua#L1296C28-L1296C28) for the grep tool subprocess, `mini.pick` could also expose the `env` argument ?

And for simplicity, it would be nice to be able to concatenate arbitrary `args` to what `mini.pick` already uses, too.

This way, we could either do:

```lua
MiniPick.builtin.grep_live({tool = "rg"}, {source = {env = {"RIPGREP_CONFIG_PATH" = "/path/to/specific/configuration"}}})
```

or

```lua
MiniPick.builtin.grep_live({tool = "rg"}, {source = {args = {"--hidden"}}})
```

This wouldn't need any tool specific handling from `mini.pick`, but would allow users to customize the calls for convenience.

**Comment by echasnovski** [comment]
> I've been wondering if there was something I could do to switch my `rg` configuration between pickers ?
> 
> For example, I'd like the default behavior of `grep_live` to use my default `rg` configuration, ignoring hidden and ignored files. But I'd also like to be able to call `grep_live` with another `rg` configuration, in case I need to find something in a hidden or ignored file.
> 
> To keep the configuration method consistent with the chosen solution, in the case of `rg` this would mean the ability, from `mini.pick`, to override `RIPGREP_CONFIG_PATH` for a given picker.

I think the best solution here would be to write a wrapper function which temporarily sets the environment variables. Something like this:
```lua
MiniPick.registry.grep_live_specific = function()
  local cache_rg_config = vim.uv.os_getenv('RIPGREP_CONFIG_PATH')
  vim.uv.os_setenv('RIPGREP_CONFIG_PATH', '/path/to/specific/configuration')
  MiniPick.builtin.grep_live({ tool = 'rg' })
  vim.uv.os_setenv('RIPGREP_CONFIG_PATH', cache_rg_config)
end
```

You can then call it with `:Pick grep_live_specific`.

> I saw that there is already support for [cwd](https://github.com/echasnovski/mini.nvim/blob/f2b89efbbad1943657e43f474fe308fceb63597e/lua/mini/pick.lua#L1296C28-L1296C28) for the grep tool subprocess, `mini.pick` could also expose the `env` argument ?
> 
> And for simplicity, it would be nice to be able to concatenate arbitrary `args` to what `mini.pick` already uses, too.

`grep_live` is a very unusual picker which doesn't operate as others. So having it affect the overall picker source design wouldn't be wise.

For other CLI related tasks there is a `builtin.cli()` picker which has `spawn_opts` local option to set anything `vim.uv.spawn()` allows as option.

**Comment by EddieFAF** [comment]
I really like your work. My setup contains only mini.nvim and a few others for LSP. But I'm having trouble getting mini.pick to work.
On 2 out of 3 systems, I only get the floating window with no content. I can ESC to close that window, but nothing else happens. That's independent of the picker I choose, neither files nor buffers shows anything. On the third system, everything works as intended. I'm using exactly the same config (same init.lua file) on all three systems.
Any hints what might causing this? :checkhealth on all three shows nothing obviously wrong.

**Comment by echasnovski** [comment]
> I really like your work. My setup contains only mini.nvim and a few others for LSP. But I'm having trouble getting mini.pick to work. On 2 out of 3 systems, I only get the floating window with no content. I can ESC to close that window, but nothing else happens. That's independent of the picker I choose, neither files nor buffers shows anything. On the third system, everything works as intended. I'm using exactly the same config (same init.lua file) on all three systems. Any hints what might causing this? :checkhealth on all three shows nothing obviously wrong.

Hi, thanks!

Without additional information I can not really help. For example:
- What is the working system and what are two that don't work?
- Does any built-in picker work? For example, `builtin.help`?
- Does a direct picker start work? Try executing `:lua require('mini.pick').start({ source = { items = { 'a', 'b' } } })` which should result into a picker with two items "a" and "b".

**Comment by EddieFAF** [comment]
> 
> Without additional information I can not really help. For example:
> 
> * What is the working system and what are two that don't work?

working: Arch, neovim 0.9.4
nonworking: debian 12, neovim 0.10.0 and debian 11, neovim 0.10.0
> * Does any built-in picker work? For example, `builtin.help`?

I'm using those via keybinding, but nothing shows beside the empty window :(
> * Does a direct picker start work? Try executing `:lua require('mini.pick').start({ source = { items = { 'a', 'b' } } })` which should result into a picker with two items "a" and "b".

Nothing shown

**Comment by echasnovski** [comment]
> working: Arch, neovim 0.9.4
> nonworking: debian 12, neovim 0.10.0 and debian 11, neovim 0.10.0

My guess would be that both 0.10.0 versions are too old and don't have [the floating window footer feature](https://github.com/neovim/neovim/pull/24739). Either using latest Nightly appimage or indeed rebuilding latest nightly from source should solve the issue in this case.

**Comment by EddieFAF** [comment]
> > working: Arch, neovim 0.9.4
> > nonworking: debian 12, neovim 0.10.0 and debian 11, neovim 0.10.0
> 
> My guess would be that both 0.10.0 versions are too old and don't have [the floating window footer feature](https://github.com/neovim/neovim/pull/24739). Either using latest Nightly appimage or indeed rebuilding latest nightly from source should solve the issue in this case.

As usual, you are absolutly right :) Rebuild both nvim versions solved my problem. Thanks.

**Comment by juantascon** [comment]
Is there a way to resume just one individual picker? I would like to set a keymap to always open live_grep with the last searched word, basically something similar to fzf live_grep_resume.

**Comment by echasnovski** [comment]
> Is there a way to resume just one individual picker? I would like to set a keymap to always open live_grep with the last searched word, basically something similar to fzf live_grep_resume.

Unfortunately, no. It would require storing too much data, so there is only one `builtin.resume()` to resume the latest used picker.

You can kind of implement this by overriding `choose` action to store the latest query and then implement other picker to use that query as initial. Something like this:
```lua
local latest_grep_live_query = {}
MiniPick.registry.grep_live = function(local_opts)
  local cache_query_and_choose = function(...)
    latest_grep_live_query = MiniPick.get_picker_query()
    return MiniPick.default_choose(...)
  end

  return MiniPick.builtin.grep_live(local_opts, { source = { choose = cache_query_and_choose } })
end

MiniPick.registry.grep_live_resume = function(local_opts)
  vim.schedule(function() MiniPick.set_picker_query(latest_grep_live_query) end)
  return MiniPick.registry.grep_live(local_opts)
end
```

With this you can use `:Pick grep_live_resume` to resume latest search. Note that it will resume with the query when the latest chosen item was chosen. So if you looked for something and hit `<Esc>`, it doesn't store the query. You can also override `<Esc>` to also store it, but I would like to leave to actual users.

**Comment by echasnovski** [comment]
With the release of 0.11.0, 'mini.pick' is now out of beta-testing. Huge thanks to everyone for constructive feedback and participation!

**Comment by marcusandre** [comment]
@echasnovski Thank you for your work and dedication! I immediately switched to it and I don't see any point to switch back.

**Comment by endir** [comment]
Thank you for the very nice plugin. Is there a way to have ignorecase when doing grep_live?

**Comment by echasnovski** [comment]
> Thank you for the very nice plugin. Is there a way to have ignorecase when doing grep_live?

`builtin.grep_live` uses CLI tool with only a minimal set of arguments to make it work. To configure it further, [configure corresponding tool](https://github.com/echasnovski/mini.nvim/blob/05f4a49cd85a67b90328a1bcbae4d9ed2a0a417b/doc/mini-pick.txt#L79-L84). If you use `rg`, then it is a matter of putting a `--ignore-case` line in some file and setting `RIPGREP_CONFIG_PATH` environment variable to point to it. See [this part of `rg` guide](https://github.com/BurntSushi/ripgrep/blob/master/GUIDE.md#configuration-file).

**Comment by endir** [comment]
Thanks a lot Evgeni! Did not know about ripgrep's config file. Works beautifully!

**Comment by urosmm** [comment]
Hi, thanks for the great plugin!
How would I go about configuring a grep and grep_live picker that would use GNU grep?

**Comment by echasnovski** [comment]
> Hi, thanks for the great plugin! How would I go about configuring a grep and grep_live picker that would use GNU grep?

`builtin.grep_live` only supports `rg` or `git` (which is `git grep` and very close to GNU grep) tools. I'd recommend using `git` tool with something like `:Pick grep_live tool='git'`.

`builtin.grep` also supports `git` tool, but if you really need GNU, you can use `builtin.cli()` picker to call `grep` CLI tool in a way similar to the output of `git grep`.

**Comment by urosmm** [comment]
Thanks for the quick reply!

It's just that I don't want to have a dependency on `rg`. Using `git` is acceptable, but is it possible to pass additional flags like `--untracked` to `git grep` for grep_live picker?

**Comment by echasnovski** [comment]
> Thanks for the quick reply!
> 
> It's just that I don't want to have a dependency on `rg`. Using `git` is acceptable, but is it possible to pass additional flags like `--untracked` to `git grep` for grep_live picker?

As `grep_live` is a fairly unique picker in how it works, there are some limitations as to it. The suggested approach is to [configure each CLI tool separately](https://github.com/echasnovski/mini.nvim/blob/baf17e6472b0968661dce5daa372251e934da39f/doc/mini-pick.txt#L79-L84), but I am not sure if `git` can be configured to use `--untracked` for only `git grep` command.

My suggestion would be to either use `builtin.cli()` with desired flags (but no "live" behavior) or use `rg`.

**Comment by GitMurf** [comment]
> [!NOTE]
> @echasnovski I'm not sure if this is appropriate to add here on this older gh issue but thought it was better than opening a new gh issue or something (since it is not an issue but a question / request). Please let me know for the future what protocol you would prefer I follow for items like these!

For mini pickers, is there a way to have preview split out to side of the main picker items? similar to how telescope does it so you can cycle through the items in the picker and have the preview shown on the right side for example. Also similar to how you have mini.files working in concept where you can navigate the list of items and the preview is shown to the right.

Thanks!

**Comment by echasnovski** [comment]
> Please let me know for the future what protocol you would prefer I follow for items like these!

There is a [separate Q&A category for discussion](https://github.com/echasnovski/mini.nvim/discussions/categories/q-a) which is dedicated for this type of interaction.

> For mini pickers, is there a way to have preview split out to side of the main picker items? similar to how telescope does it so you can cycle through the items in the picker and have the preview shown on the right side for example. Also similar to how you have mini.files working in concept where you can navigate the list of items and the preview is shown to the right.

I am afraid, no. The whole idea is to have it be single window only while separate preview needs extra window. The 'mini.pick' approach is to have preview enabled with `<Tab>` and then navigate up/down: this will stay in preview mode.

**Comment by GitMurf** [comment]
> I am afraid, no. The whole idea is to have it be single window only while separate preview needs extra window. The 'mini.pick' approach is to have preview enabled with `<Tab>` and then navigate up/down: this will stay in preview mode.

No worries! Thanks for the response. I kind of figured as such given your approach (which I appreciate) with trying to keep code as simple and concise as possible while maximizing features/value. Cycling through while in preview gets me most the way there.

**Comment by drowning-cat** [comment]
I would add something like this:

```lua
MiniPick.builtin.files({ args = { fd = { --[[ List of arguments ]] } } })
```

**Comment by echasnovski** [comment]
> I would add something like this:

If you mean to suggest a way to pass CLI tool arguments to some pickers, then this was suggested in #830.

**Comment by drowning-cat** [comment]
> If you mean to suggest a way to pass CLI tool arguments to some pickers, then this was suggested in [#830](https://github.com/nvim-mini/mini.nvim/issues/830).

I’ve got doubts about my search skills. Thanks ;0
